File: frontend/src/types/oauth.ts
/**
 * OAuth 三方登录 —— 前端类型定义（T04）
 * 契约来源：docs/design-oauth-login.md §2.3（接口契约固化，勿随意改动字段名）
 */

/** ticket 状态（后端 oauth_tickets.status，§2.1.3 / §2.3.2 接口 4） */
export type OAuthTicketStatus =
  | 'LOGIN_OK'
  | 'BIND_REQUIRED'
  | 'REGISTER_REQUIRED'
  | 'CONFIRM_REQUIRED'
  | 'ALREADY_BOUND'
  | 'BIND_CONFLICT'

/** 授权意图：login = 登录/注册；bind = 已登录加绑 */
export type OAuthIntent = 'login' | 'bind'

/** 客户端类型：web 浏览器 302 回调 / desktop Electron 本地回环（T05 适配） */
export type OAuthClientType = 'web' | 'desktop'

/** 接口 1：GET /auth/oauth/providers 返回的 provider 条目 */
export type OAuthProviderInfo = {
  name: string
  display_name: string
  authorize_path: string
  icon_key: string
}

export type OAuthProvidersResponse = {
  providers: OAuthProviderInfo[]
}

/** 接口 2：GET /auth/oauth/{provider}/authorize 响应 */
export type OAuthAuthorizeResponse = {
  authorize_url: string
  state: string
  expires_in: number
}

export type OAuthAuthorizeParams = {
  intent?: OAuthIntent
  client_type?: OAuthClientType
  /** 授权完成后前端应落地的站内相对路径（后端校验，禁止绝对 URL） */
  redirect_after?: string
}

/** TokenResponse（路径 A / B / C 成功后签发） */
export type OAuthTokenResponse = {
  access_token: string
  refresh_token: string
  token_type: string
}

/** 接口 4：POST /auth/oauth/resolve 响应（幂等读，不消费 ticket） */
export type OAuthResolveResult = {
  status: OAuthTicketStatus
  provider?: string
} & Partial<OAuthTokenResponse> & {
    /** status=BIND_REQUIRED：🔴 脱敏邮箱（如 z***@example.com），未认证端点不会返回完整邮箱 */
    email_masked?: string
    email_verified?: boolean
    /** status=REGISTER_REQUIRED：三方 email 仅作预填默认值，以用户手填为准（拍板 #6） */
    suggested_email?: string
    suggested_display_name?: string
    suggested_avatar_url?: string
    /** status=CONFIRM_REQUIRED：管理员账号加绑需二次密码确认 */
    reason?: string
    provider_display_name?: string
    /** status=ALREADY_BOUND */
    bound_at?: string
  }

/** 接口 5：POST /auth/oauth/bind 响应（已登录加绑终态） */
export type OAuthBindResult = {
  status: 'BOUND' | 'ALREADY_BOUND'
  identity?: OAuthIdentity
}

/** 接口 8：GET /auth/oauth/identities 返回的单条绑定身份 */
export type OAuthIdentity = {
  id: string
  provider: string
  provider_display_name: string
  provider_email: string
  provider_avatar_url: string
  created_at: string
  last_login_at: string
}

export type OAuthIdentitiesResponse = {
  identities: OAuthIdentity[]
  available_providers: string[]
}

/** 接口 6/7 终态成功响应：绑定/注册完成并签发 token */
export type OAuthCompleteResponse = {
  status: 'BOUND' | 'REGISTERED'
  access_token: string
  refresh_token: string
  token_type: string
}

/** 后端 OAuthAPIError 统一错误体（§4.5）：{ detail, code } */
export type OAuthErrorBody = {
  detail?: string
  code?: string
  /** OAUTH_TICKET_LOCKED（423）时的冷却秒数 */
  retry_after?: number
  /** OAUTH_PASSWORD_INVALID（401）时后端可附带的剩余尝试次数（E-18） */
  attempts_left?: number
}

File: backend/app/domains/auth/schemas/oauth.py
"""
OAuth 三方登录 Pydantic Schemas（对应设计文档 §2.3 接口契约）。

所有 schema 仅做数据搬运与基础校验；账号判定逻辑集中在
``services/oauth_service.py``（T02），不得前移到 schema 层。
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, EmailStr, Field

from app.domains.auth.schemas.auth import TokenResponse

# ── ticket status 常量（与 models/oauth.py 保持同一组值；schema 层引用避免循环导入）──
from app.domains.auth.models.oauth import (  # noqa: F401
    TICKET_STATUS_ALREADY_BOUND,
    TICKET_STATUS_BIND_CONFLICT,
    TICKET_STATUS_BIND_REQUIRED,
    TICKET_STATUS_CONFIRM_REQUIRED,
    TICKET_STATUS_LOGIN_OK,
    TICKET_STATUS_REGISTER_REQUIRED,
)


# ══════════════════ 接口 1：GET /auth/oauth/providers ══════════════════

class ProviderInfo(BaseModel):
    """已启用 provider 的展示信息（登录页动态渲染用）。"""

    name: str
    display_name: str
    authorize_path: str
    icon_key: str


class ProviderListResponse(BaseModel):
    """未配置 client_id 的 provider 不会出现在列表中（NFR-M2）；空列表时前端隐藏三方按钮区。"""

    providers: list[ProviderInfo]


# ══════════════════ 接口 2：GET /auth/oauth/{provider}/authorize ══════════════════

class AuthorizeParams(BaseModel):
    """authorize 端点的 query 参数（FastAPI 以 Depends 形式消费）。"""

    intent: str = Field(default="login", pattern="^(login|bind)$")
    client_type: str = Field(default="web", pattern="^(web|desktop)$")
    # 授权完成后前端应落地的站内相对路径；禁止 // 与绝对 URL（防开放重定向，router 层校验）
    redirect_after: Optional[str] = None


class AuthorizeResponse(BaseModel):
    """前端拿到 authorize_url 后 window.location.href 跳转；state 校验是服务端职责。"""

    authorize_url: str
    state: str
    expires_in: int


# ══════════════════ 接口 4：POST /auth/oauth/resolve（幂等读） ══════════════════

class TicketResolveRequest(BaseModel):
    ticket: str = Field(..., min_length=1, max_length=64)


class ResolveResponse(BaseModel):
    """按 status 分支的统一响应体。

    - LOGIN_OK：access_token / refresh_token / token_type 有值
    - BIND_REQUIRED：🔴 只回脱敏邮箱 email_masked，严禁完整邮箱（AC-S7 / NFR-S7）
    - REGISTER_REQUIRED：suggested_email / suggested_display_name / suggested_avatar_url 预填
    - CONFIRM_REQUIRED：reason="admin_bind"
    - ALREADY_BOUND：bound_at
    - BIND_CONFLICT：仅 provider
    """

    status: str
    provider: Optional[str] = None
    email_masked: Optional[str] = None
    suggested_email: Optional[str] = None
    suggested_display_name: Optional[str] = None
    suggested_avatar_url: Optional[str] = None
    email_verified: Optional[bool] = None
    # CONFIRM_REQUIRED（管理员加绑）：reason="admin_bind" + provider 展示名（§2.3 接口 4 契约）
    provider_display_name: Optional[str] = None
    reason: Optional[str] = None
    bound_at: Optional[datetime] = None
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    token_type: Optional[str] = None


# ══════════════════ 接口 5/6/7：终态请求体 ══════════════════

class OAuthBindRequest(BaseModel):
    """加绑终态（已登录态，设置页）。管理员账号加绑必须传 password（拍板 #8）。"""

    ticket: str = Field(..., min_length=1, max_length=64)
    password: Optional[str] = Field(default=None, max_length=128)


class OAuthBindConfirmRequest(BaseModel):
    """路径 B 确认绑定（未登录态）。🔴 必须验证密码，错误响应与"账号不存在"不可区分。"""

    ticket: str = Field(..., min_length=1, max_length=64)
    password: str = Field(..., min_length=1, max_length=128)


class OAuthRegisterRequest(BaseModel):
    """路径 C 补全注册。🔴 手填优先（拍板 #6）：email 以用户手填为准建号。

    校验顺序见 §2.3 接口 7：原子消费 ticket → status 校验 → normalize →
    格式 → 域名白名单 → 密码长度 → 邮箱唯一性 → 建号绑定。
    """

    ticket: str = Field(..., min_length=1, max_length=64)
    email: EmailStr
    password: str = Field(..., min_length=6, max_length=128)  # 与 UserRegister 一致
    display_name: str = Field(..., min_length=1, max_length=100)


# ══════════════════ 接口 5/6/7：终态响应体 ══════════════════

class BindResultResponse(BaseModel):
    """加绑成功响应。"""

    status: str = "BOUND"
    identity: "OAuthIdentityResponse"


class BindConfirmResponse(TokenResponse):
    """路径 B 成功：绑定 + 登录（含 token 对）。"""

    status: str = "BOUND"


class OAuthRegisterResponse(TokenResponse):
    """路径 C 成功：建号 + 绑定 + 登录（含 token 对）。"""

    status: str = "REGISTERED"


# ══════════════════ 接口 8/9：身份列表与解绑 ══════════════════

class OAuthIdentityResponse(BaseModel):
    """已绑定身份（设置页展示）。不暴露 provider_uid 明文之外的敏感字段。"""

    id: str
    provider: str
    provider_display_name: Optional[str] = None
    provider_email: Optional[str] = None
    provider_avatar_url: Optional[str] = None
    email_verified: Optional[bool] = None
    created_at: datetime
    last_login_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class IdentityListResponse(BaseModel):
    """当前用户已绑定身份列表 + 当前启用的 provider（前端渲染"可绑定"按钮）。"""

    identities: list[OAuthIdentityResponse]
    available_providers: list[str]


# BindResultResponse 对 OAuthIdentityResponse 的前向引用在此解析
BindResultResponse.model_rebuild()
